File: 198.House_Robber.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

class Solution(object):
    def rob(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """
        if nums == []:
            return 0

        maxs = {}
        for i, value in enumerate(nums):
            maxs[i] = max(value + maxs.get(i-2, 0), maxs.get(i-1, 0))

        return maxs[len(nums)-1]

class Solution:
    def rob(self, nums: List[int]) -> int:
        """
        More concise solution that uses less memory:

        O(N) Time
        O(1) Space
        """
        prev = curr = 0
        for n in nums:
            prev, curr = curr, max(n+prev, curr)
        return curr

File: 198.House_Robber.go
package main

func rob(nums []int) int {

	if len(nums) == 0 {
		return 0
	}

	maxs := map[int]int{}

	for i, value := range nums {

		var p1_max, p2_max int

		if val, ok := maxs[i-1]; ok {
			p1_max = val
		} else {
			p1_max = 0
		}

		if val, ok := maxs[i-2]; ok {
			p2_max = val
		} else {
			p2_max = 0
		}

		maxs[i] = max(value+p2_max, p1_max)
		fmt.Println(i, maxs)
	}

	return maxs[len(nums)-1]
}

func max(n, m int) int {
	if m > n {
		return m
	}
	return n
}
